util/UnitEvent.lua

```diff
@@ -0,0 +1,186 @@
+--[[
+    UnitEvent (Namesake)
+        Originally by Bribe
+
+    Uses Magic Defense instead of Defend ability to detect leave.    
+]]
+
+require('CeresStdLib.base.Rawcode')
+require('CeresStdLib.base.Optimizations')
+require('CeresStdLib.base.Native')
+require('CeresStdLib.base.Init')
+require('CeresStdLib.base.Log')
+
+require('CeresStdLib.util.World')
+
+DETECT_LEAVE            = fromRawCode('Amdf')
+
+EVENT_UNIT_ENTER        = 1
+EVENT_UNIT_LEAVE        = 2
+
+UnitEvent = {}
+
+local trigTable         = {}
+local __userData        = {}
+__userData[-1]          = {}   --  Holds references to each unit
+__userData[-2]          = {}   --  Holds preplaced flag
+__userData[0]           = 0
+
+trigTable[3], trigTable[4]                  = {}, {}
+
+local __evType          = 0
+local __evData          = {0, nil}
+local __initialized     = false
+
+replaceNative('GetUnitUserData', function(u)
+    return __userData[u]
+end)
+replaceNative('SetUnitUserData', function(u, newField)
+    __userData[u] = newField
+end)
+
+function __userData:alloc()
+    local i = __userData[0]
+    if i == 0 then
+        i                   = i + 1
+        __userData[0]       = i
+        return i
+    end
+    if (not __userData[i]) or (__userData[i] == 0) then
+        __userData[0]     = i + 1
+        __userData[i]     = -1
+        i                 = __userData[0]
+    else
+        __userData[0]     = __userData[i]
+        __userData[i]     = -1
+    end
+    return i
+end
+
+function __userData:dealloc(i)
+    if (not i) or (__userData[i] ~= -1) then return end
+
+    __userData[i], __userData[0]         = __userData[0], i
+end
+
+function GetUnitId(u) return Native.GetUnitUserData(u) end

```

Comment:
Do we have a reason to have recyclable IDs now? They were implemented as a necessary evil to be able to fit unit-unique data into strictly sized JASS arrays, but now we only have maps, so HandleId do the job probably even better.

Reply:
They were included out of habit when I first wrote it up. Would the best course of action be to remove them?